### PR TITLE
OpenShift permissions

### PR DESCRIPTION
**Pull request by pohly**
This consists of two changes:
- running the controller without a fixed UID to allow OpenShift to assign a UID dynamically (see commit for a detailed discussion)
- adding one additional role binding for the node driver

Fixes: https://github.com/intel/pmem-csi/issues/935

### CODE REVIEW

File: deploy/common/kata-app.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+kind: Pod
```
Comment (avalluri) [inline deploy/common/kata-app.yaml:1]: @pohly Is this change related to this commit?

File: deploy/kustomize/driver/pmem-csi.yaml [hunk 2]
```diff
@@ -147,10 +147,6 @@ spec:
         app.kubernetes.io/instance: pmem-csi.intel.com
         pmem-csi.intel.com/webhook: ignore
     spec:
-      securityContext:
-        runAsNonRoot: true
```
Comment (avalluri) [inline deploy/kustomize/driver/pmem-csi.yaml:151]: Does running the controller as non-root(though not choosing the UID) is still not relevant?

File: deploy/olm-catalog/0.9.0/pmem-csi-operator_v1_serviceaccount.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+apiVersion: v1
```
Comment (avalluri) [inline deploy/olm-catalog/0.9.0/pmem-csi-operator_v1_serviceaccount.yaml:1]: I think this is not intentional to add files under 'deploy/olm-catalog', is it?

File: deploy/kustomize/driver/pmem-csi.yaml [hunk 4]
```diff
@@ -4,6 +4,26 @@ metadata:
   name: pmem-csi-intel-com-controller
   namespace: default
 ---
+# This role binding corresponds to what the following command creates:
```
Comment (avalluri) [inline deploy/kustomize/driver/pmem-csi.yaml:7]: Isn't it good to keep these kinds of changes in a kustomize layer specific to openshift?

**Review by avalluri — COMMENTED** [review]
> On non-OpenShift systems the additional RoleBinding is ignored,

But it might raise unnecessary confusion for the cluster-admin.

File: deploy/common/kata-app.yaml (hunk 1 above)
Comment (pohly) [reply deploy/common/kata-app.yaml:1]: No, will remove.

File: deploy/kustomize/driver/pmem-csi.yaml (hunk 2 above)
Comment (pohly) [reply deploy/kustomize/driver/pmem-csi.yaml:151]: Have you seen the commit message? There I am trying to discuss that aspect.

In a nutshell, actual users will still run this as non-root.

File: deploy/olm-catalog/0.9.0/pmem-csi-operator_v1_serviceaccount.yaml (hunk 3 above)
Comment (pohly) [reply deploy/olm-catalog/0.9.0/pmem-csi-operator_v1_serviceaccount.yaml:1]: Not intentional. `.gitignore` should have been updating when adding that directory to the built files.

File: deploy/kustomize/driver/pmem-csi.yaml (hunk 4 above)
Comment (pohly) [reply deploy/kustomize/driver/pmem-csi.yaml:7]: That would imply that we maintain different deployments and deployment instructions for upstream Kubernetes and OpenShift, with the resulting user errors when they pick the wrong one.

I'm trying to avoid that.

**Comment by pohly** [comment]
> > On non-OpenShift systems the additional RoleBinding is ignored,
> But it might raise unnecessary confusion for the cluster-admin.

Only if they run on upstream Kubernetes *and* inspect role bindings. I think this is very unlikely to be a problem.

File: deploy/olm-catalog/0.9.0/pmem-csi-operator_v1_serviceaccount.yaml (hunk 3 above)
Comment (pohly) [reply deploy/olm-catalog/0.9.0/pmem-csi-operator_v1_serviceaccount.yaml:1]: Ah, it was the other way around: I still had the directory, but 435f821906c055b8dfd6a918ba36dc3e2137dcea removed the exclude rule.

File: deploy/kustomize/driver/pmem-csi.yaml (hunk 2 above)
Comment (pohly) [reply deploy/kustomize/driver/pmem-csi.yaml:151]: docs/design.md also gets updated to explain this.

File: deploy/kustomize/testing/controller-coverage-patch.yaml [hunk 5]
```diff
@@ -14,23 +14,3 @@
     hostPath:
       path: /var/lib/pmem-csi-coverage
       type: DirectoryOrCreate
-# driver in controller mode runs as non-root user
-# which cannot have enough permissions to access
-# coverage-dir(/var/lib/pmem-csi-coverage)
-# Change the ownership of this folder to "pmem-csi:pmem-csi"
-# so that driver could write coverage reports.
-- op: add
-  path: /spec/template/spec/initContainers
```
Comment (avalluri) [inline deploy/kustomize/testing/controller-coverage-patch.yaml:23]: Seems like this change(removing init container) is not related to this commit?

File: deploy/kustomize/driver/pmem-csi.yaml (hunk 2 above)
Comment (avalluri) [reply deploy/kustomize/driver/pmem-csi.yaml:151]: My point is can't we keep `runAsNotRoot: true` by not setting the UID?

File: deploy/kustomize/testing/controller-coverage-patch.yaml (hunk 5 above)
Comment (pohly) [reply deploy/kustomize/testing/controller-coverage-patch.yaml:23]: It is. Because we now run as root (at least where we test), we don't need this workaround for non-root anymore.

File: deploy/kustomize/driver/pmem-csi.yaml (hunk 2 above)
Comment (pohly) [reply deploy/kustomize/driver/pmem-csi.yaml:151]: But on upstream Kubernetes we do run as root. We can't keep `runAsNotRoot: true` unless we set the fixed UID, can we?

**Comment by pohly** [comment]
@avalluri did I address your comments?

**Comment by pohly** [comment]
Also note the new final commit `deploy: test with "testing" deployments instead of "coverage"`. That was what I found when trying out this PR on an OpenShift 4.6 cluster. The Pods didn't come up because image pulling in that cluster is impossible (access to Internet not configured), but at least the Pods all got created.

**Review by avalluri — APPROVED** [review]

**Comment by pohly** [comment]
I have rebased and tested the rebased deployment on an OpenShift 4.8 (in development!) cluster that was created with libvirt such that the worker node has 896.00 MiB of PMEM.

```
test/setup-ca-kubernetes.sh
sed -e 's;intel/pmem-csi-driver:canary;pohly/pmem-csi-driver:canary;' deploy/kubernetes-1.21/pmem-csi-direct.yaml | kubectl apply -f -
kubectl label node/localhost storage=pmem
```

The "localhost" name is a quirk of the cluster - somehow the worker node lost its real name.

All pods come up and PMEM-CSI is able to find the PMEM:
```
I0623 15:07:43.865223       1 main.go:73] "PMEM-CSI started." version="v0.9.0-197-g157b2c818"
I0623 15:07:43.865516       1 mount_linux.go:163] Detected OS without systemd
I0623 15:07:43.867050       1 mount_linux.go:163] Detected OS without systemd
I0623 15:07:43.867720       1 server.go:54] "GRPC-server: Listening for connections" endpoint="unix:///csi/csi.sock"
I0623 15:07:43.868426       1 pmem-csi-driver.go:305] "PMEM-CSI ready." capacity="896Mi maximum volume size, 896Mi available, 896Mi managed, 896Mi total"
```

There was an issue with creating a 100Mi namespace in direct mode, but LVM mode worked. That's unrelated to this PR, so I'll merge it.
